csrc/id_model/id_model.h

```diff
@@ -192,7 +192,35 @@ class IdModel : public PolymorphicBase {
   // input is promoted, the output needs to be promoted too. If
   // there's already an equivalent expr that uses the promoted inputs,
   // create a mapping from the outputs of the IEL expr to the outputs
-  // of the equivalent expr.
+  // of the equivalent expr. When require_loop_mapped_promotion is
+  // true, the equivalent expr needs to be already loop mapped. If no
+  // such expr is found, the IEL expr is replayed iwth the promoted
+  // inputs. require_loop_mapped_promotion is true when this function
+  // is used for step 3.
+  //
+  // This is used twice when building the promotion map. The first time
+  // it is used there's no loop graph promotion yet, so only the IEL
+  // promotions are propagated. In that case, loop_graph_promotion_map
+  // should be just empty.
+  //
+  // Propagation uses iel_promotion_map and
+  // loop_graph_promotion_map. If both are available for an IEL group,
+  // the former has the precedence. This is because when this function
+  // is used for step 4, the given iel_promotion_map starts as an
+  // empty map and gets populated during this propagation, so any
+  // mapping in the map is guaranteed to be the correct final mapping,
+  // whereas the loop graph may have invalid mappings for partially
+  // inlined domains.
+  void propagatePromotionsInIELGraph(
+      const ValGraph& iel_graph,
+      std::unordered_map<ValGroup, IterDomain*>& iel_promotion_map,
+      const ValGraph& loop_graph,
```

Comment:
Is this parameter needed? Can't we just call `idGraph(IdMappingMode::LOOP)` inside this function?

Reply:
Generally, yes, but may not be necessary for our use cases. Note that `loop_promotion_map` may have been built with an old version of `idGraph(IdMappingMode::LOOP)`. There's no guarantee that the current loop graph, `idGraph(IdMappingMode::LOOP)`, is the same as the graph used to build `loop_promotion_map`. Replaying domains could create a new val group, and if it's merged with an existing group, the domains in the existing group would be moved to a new group. Replay should not change the logical structure of the graph, but the pointer values of val groups could change, which means querying with `loop_promotion_map` using the current loop graph could fail if the map was originally built with an old version of the loop graph.

I'll file an issue and revisit this later, probably with the final Step 5 PR.

Reply:
#2029